### PR TITLE
docs(user-feedback): implemented structured feedback for the processes

### LINKED ISSUE #29: Crawler: Print summary of phases
@NOTE => Up to date until here: https://github.com/timdeschryver/deprecation-manager/issues/29#issuecomment-667708124

As we have Processes and or tasks (we should define that) we should define when we give feedback.
I imagine it as process lifecycles.
A minimal set of hooks would be start and end. 

I suggest feedback in for 2 hooks. 
- start
- end

The feedback gives a full overview of the process. It lists every major and minor feedback. Furthermore, this version should serve as documentation, help, and guide through the step and explains a little bit of what happened or what will change in future runs.

An interface that specifies that could look like the following: 
```ts
/**
 * @description
 * This interface is here to uniform feedback form processes.
 * It specifies a minimal and a detail-level of feedback (verbose option in mind).
 *
 * Every process
 */
export interface ProcessFeedback {
  printStart(rawRelease: CrawledRelease, config: CrawlConfig): void
  printEnd(rawRelease: CrawledRelease, config: CrawlConfig): void
}
```

In future releases we could also think about displaying the progress:

```ts
export interface AsyncProcessFeedback extends ProcessFeedback {
  printProgress(rawRelease: CrawledRelease, config: CrawlConfig): void
}
```

## Setup Phase

-  [x] printStart
```
= SETUP PHASE ====================================

🌱 Start setting up crawler.config.json

=== ⏰

```
-  [x] printEnd
```

✓ Repository configured for deprecation management in brach/tag ${gitTag} done!

Configuration saved under: ${crawlerConfigPath}

Output Directory: ${outputDir}
Typescript Configfile: ${tsConfigPath}
Deprecation Keyword: @deprecated
Link to Information: ${link}#ruid

From now on the crawler will go to deprecation-crawler.config.json and crawl files referenced under tsconfig.deprecation-crawler.ts and this questions will not get asked next time.

If you want to change something edit the content of deprecation-crawler.config.json or tsconfig.deprecation-crawler.ts. 

===============================================
```

### Full Sample:

```
= SETUP PHASE ====================================

🌱 Start setting up crawler.config.json

=== ⏰

✓ Repository configured for deprecation management in brach/tag ${gitTag} done!

Configuration saved under: ${crawlerConfigPath}

Output Directory: ${outputDir}
Typescript Configfile: ${tsConfigPath}
Deprecation Keyword: @deprecated
Link to Information: ${link}#ruid

From now on the crawler will go to deprecation-crawler.config.json and crawl files referenced under tsconfig.deprecation-crawler.ts and these questions will not get asked next time.

If you want to change something edit the content of deprecation-crawler.config.json or tsconfig.deprecation-crawler.ts. 

===============================================
```

## Crawl Phase

-  [x] printStart
```
= CRAWL PHASE ====================================

🔎 Looking for deprecations in in branch/tag ${config.gitTag}.

Date: ${config.gitTagDate}

=== ⏰

```

-  [x] printEnd
```

✓ Found ${numDeprecations} in ${numFiles} files.

Files: ${numFiles}
Deprecations: ${numDeprecations}

===============================================
```

### Full Sample:
```
= CRAWL PHASE ====================================

🔎 Looking for deprecations in in branch/tag ${config.gitTag}.

Date: ${config.gitTagDate

=== ⏰

✓  Found ${numDeprecations} in ${numFiles} files.

Files: ${numFiles}
Deprecations: ${numDeprecations}

===============================================
```

## Repo Sync Phase

 -  [x] printStart
```
= REPOSITORY SYNC PHASE =======================

💾 Start syncing crawled result to repository ${gitTag}

=== ⏰

```

 -  [x] printEnd
```

✓  Repository synced! 
Added ${numDeprecations} links to deprecations in ${numGroups} files.

- 3 in test-folder-for-crawling\multiple-string-patterns-at-once.test.ts
- 3 in test-folder-for-crawling\all-lowercase.test.ts
- 3 in test-folder-for-crawling\whitespace-normalisation.test.ts
=====================================================
```

Full Sample:
```
= REPOSITORY SYNC PHASE =======================

💾 Start syncing crawled result to repository ${gitTag}

=== ⏰

✓  Repository synced! 
Added ${numDeprecations} links to deprecations in ${numGroups} files.

- 3 in test-folder-for-crawling\multiple-string-patterns-at-once.test.ts
- 3 in test-folder-for-crawling\all-lowercase.test.ts
- 3 in test-folder-for-crawling\whitespace-normalisation.test.ts
=====================================================
```

## Grouping Phase

 - [x] printStart
```
= GROUPING PHASE ====================================
⚙️ Start grouping of deprecations. 

=== ⏰

```

 - [x] printEnd
```
✓  Grouping is done! 
Grouped ${numAllDeprecations} deprecations in ${numAllGroups} files.

Deprecations: ${numDeprecations}
Groups: ${numGroups}
- ungrouped: ${numGroupDeprecations} (${numMatchers} matcher)
- internal: ${numGroupDeprecations} (${numMatchers} matcher)
- multi: ${numGroupDeprecations} (${numMatchers} matcher)
- whitespace: ${numGroupDeprecations} (${numMatchers} matcher)

===

Group: ${groupName}
Matchers: ${numMatchers}
Deprecations: ${numDeprecations}

================================
```

### Full Sample

```
= GROUPING PHASE ====================================
⚙️ Start grouping of deprecations. 

=== ⏰

✓  Grouping is done! 
Grouped ${numAllDeprecations} deprecations in ${numAllGroups} files.

Deprecations: ${numDeprecations}
Groups: ${numGroups}
- ungrouped: ${numGroupDeprecations} (${numMatchers} matcher)
- internal: ${numGroupDeprecations} (${numMatchers} matcher)
- multi: ${numGroupDeprecations} (${numMatchers} matcher)
- whitespace: ${numGroupDeprecations} (${numMatchers} matcher)
================================
```

## Update Format Phase
- [x] printStart
```
= UPDATE FORMATS PHASE - MARKDOWN ========================

✍️ Start updating the markdown output. 

=== ⏰

```

- [x] printEnd
```
✓  Markdown updated! 
Added ${numDeprecations} deprecations in ${numFiles} files.

=====================================================
```

Full Sample:
```
= UPDATE FORMATS PHASE - MARKDOWN ========================

✍️ Start updating the markdown output. 

=== ⏰

✓  Markdown updated! 
Added ${numDeprecations} deprecations in ${numFiles} files.

=====================================================
```

---

Possible implementation:
```ts
const printRawJsonSummary = (crawledRelease: CrawledRelease, config: CrawlConfig): void => {
  const numGroups: number = config.groups.length;
  const numDeprecations: number = crawledRelease.deprecations.length;
  console.log(``);
  console.log(`= GROUPING PHASE ====================================`);
  console.log(`Branch: ${crawledRelease.version} (${crawledRelease.date})`);
  console.log(`Groups: ${numGroups}`);
  console.log(`Deprecations: ${numDeprecations}`);
  console.log(`=====================================================`);
  console.log(``);
}

```

### LINKED ISSUE #79: If groupint is escaped run through all the remaining deprecations with the current groups

### PR DESCRIPTION
**Pull request by BioPhoton**
Adds feedback to:
- setup
- crawl
- grouping
- repo sync
- format
Introduces `--verbose` flag

closes #29 closes #79

### CODE REVIEW

File: packages/deprecation-crawler/tests/index.test.ts [hunk 1]
```diff
@@ -34,16 +34,18 @@ beforeAll(async () => {
...
   expect(cliOutput).toMatch(/Running with tsconfig: tsconfig.sandbox.json/i);
-  expect(cliOutput).toMatch(/Looking for deprecations/i);
-  expect(cliOutput).toMatch(/Adding ruid to deprecations/i);
-  expect(cliOutput).toMatch(/Regenerating raw JSON/i);
-  expect(cliOutput).toMatch(/Start grouping deprecations/i);
+  expect(cliOutput).toMatch(/SETUP PHASE/i);
+  expect(cliOutput).toMatch(/tsconfig.sandbox.json/i);
+  expect(cliOutput).toMatch(/CRAWL PHASE/i);
+  expect(cliOutput).toMatch(/GROUPING PHASE/i);
+  expect(cliOutput).toMatch(/FORMAT OUTPUT/i);
```
Comment (BioPhoton) [inline packages/deprecation-crawler/tests/index.test.ts:45]: You can add more detailed tests as the `--verbose` flag is added.

**Review by timdeschryver — COMMENTED** [review]
Do you think we should abstract all logging logic to the `log` module?
For example, in other files you wouldn't use `kleur`, nor `console.log`.

I've added some grammar feedback and another comment to replace static config.

File: packages/deprecation-crawler/src/lib/crawler/index.ts [hunk 2]
```diff
@@ -39,7 +40,10 @@ function crawlFileForDeprecations(
       /\\/g,
       '/'
     );
-    console.log(`🔎 Looking for deprecations in ${path}`);
+
+    if (getVerboseFlag()) {
+      console.log(`🔎 Looking for deprecations in ${path}`);
```
[the remark this answers is not in the document]
Comment (timdeschryver) [reply packages/deprecation-crawler/src/lib/crawler/index.ts:45]: Can't commit this one via github, `kleur` needs to be imported first.

**Comment by BioPhoton** [comment]
> Do you think we should abstract all logging logic to the `log` module?
> For example, in other files you wouldn't use `kleur`, nor `console.log`.
> 
> I've added some grammar feedback and another comment to replace static config.

Yes thought about it, but was too lazy :D

File: packages/deprecation-crawler/src/lib/config.ts [hunk 3]
```diff
@@ -20,5 +34,37 @@ export async function getConfig(): Promise<CrawlConfig> {
...
+    },
+
+    printInitialEnd(config: CrawlConfig): void {
+      console.log(
+        kleur.green(`✓ `),
+        kleur.gray(`Repository configured for deprecation management in done!`)
+      );
+      console.log(kleur.gray(`Configuration saved under: ${getConfigPath()}`));
+      console.log(kleur.gray(JSON.stringify(config, null, 4)));
+      console.log(`From now on the crawler will go to deprecation-crawler.config.json and crawl files referenced under tsconfig.deprecation-crawler.ts and this questions will not get asked next time.
```
[the remark this answers is not in the document]
Comment (BioPhoton) [reply packages/deprecation-crawler/src/lib/config.ts]: done

File: packages/deprecation-crawler/src/lib/crawler/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (BioPhoton) [reply packages/deprecation-crawler/src/lib/crawler/index.ts:45]: done

**Pull request merged**